File: users-service/src/main/java/net/plshark/users/service/GroupsServiceImpl.java
package net.plshark.users.service;

import javax.annotation.Nonnull;
import lombok.AllArgsConstructor;
import net.plshark.errors.DuplicateException;
import net.plshark.errors.ObjectNotFoundException;
import net.plshark.users.model.Group;
import net.plshark.users.repo.GroupRolesRepository;
import net.plshark.users.repo.GroupsRepository;
import org.springframework.dao.DataIntegrityViolationException;
import org.springframework.stereotype.Component;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

/**
 * Group management service implementation
 */
@Component
@AllArgsConstructor
public class GroupsServiceImpl implements GroupsService {

    @Nonnull
    private final GroupsRepository groupsRepo;
    @Nonnull
    private final GroupRolesRepository groupRolesRepo;

    @Override
    public Mono<Group> get(String name) {
        return groupsRepo.getForName(name);
    }

    @Override
    public Mono<Group> getRequired(String name) {
        return get(name).switchIfEmpty(Mono.error(() -> new ObjectNotFoundException("No group found for " + name)));
    }

    @Override
    public Flux<Group> getGroups(int maxResults, long offset) {
        return groupsRepo.getGroups(maxResults, offset);
    }

    @Override
    public Mono<Group> create(Group group) {
        return groupsRepo.insert(group)
                .onErrorMap(DataIntegrityViolationException.class, e -> new DuplicateException("A group with name " +
                        group.getName() + " already exists", e));
    }

    @Override
    public Mono<Void> delete(long groupId) {
        return groupsRepo.delete(groupId);
    }

    @Override
    public Mono<Void> delete(String name) {
        //noinspection ConstantConditions
        return get(name)
                .flatMap(group -> delete(group.getId()));
    }

    @Override
    public Mono<Void> addRoleToGroup(long groupId, long roleId) {
        return groupRolesRepo.insert(groupId, roleId);
    }

    @Override
    public Mono<Void> removeRoleFromGroup(long groupId, long roleId) {
        return groupRolesRepo.delete(groupId, roleId);
    }
}

File: users-service/src/main/kotlin/net/plshark/users/service/GroupsServiceImpl.kt
package net.plshark.users.service

import net.plshark.errors.DuplicateException
import net.plshark.errors.ObjectNotFoundException
import net.plshark.users.model.Group
import net.plshark.users.model.GroupCreate
import net.plshark.users.repo.GroupRolesRepository
import net.plshark.users.repo.GroupsRepository
import org.springframework.dao.DataIntegrityViolationException
import org.springframework.stereotype.Component
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono

/**
 * Group management service implementation
 */
@Component
class GroupsServiceImpl(private val groupsRepo: GroupsRepository, private val groupRolesRepo: GroupRolesRepository) : GroupsService {

    override fun get(name: String): Mono<Group> {
        return groupsRepo.getForName(name)
    }

    override fun getRequired(name: String): Mono<Group> {
        return get(name).switchIfEmpty(Mono.error { ObjectNotFoundException("No group found for $name") })
    }

    override fun getGroups(maxResults: Int, offset: Long): Flux<Group> {
        return groupsRepo.getGroups(maxResults, offset)
    }

    override fun create(group: GroupCreate): Mono<Group> {
        return groupsRepo.insert(group)
            .onErrorMap(DataIntegrityViolationException::class.java) { e: DataIntegrityViolationException? ->
                DuplicateException("A group with name ${group.name} already exists", e)
            }
    }

    override fun delete(groupId: Long): Mono<Void> {
        return groupsRepo.delete(groupId)
    }

    override fun delete(name: String): Mono<Void> {
        return get(name)
            .flatMap { group: Group -> delete(group.id) }
    }

    override fun addRoleToGroup(groupId: Long, roleId: Long): Mono<Void> {
        return groupRolesRepo.insert(groupId, roleId)
    }

    override fun removeRoleFromGroup(groupId: Long, roleId: Long): Mono<Void> {
        return groupRolesRepo.delete(groupId, roleId)
    }
}